File: src/backend/mt/impl/compute.m
/*
 * Copyright (C) 2020 Recep Aslantas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../common.h"
#include "../../../api/compute_internal.h"

typedef struct MTComputePipelineDesc {
  id<MTLFunction> function;
} MTComputePipelineDesc;

static MTComputeEncoder *
mt_computeEncoder(GPUComputePassEncoder *enc) {
  return enc ? enc->_priv : NULL;
}

static id<MTLBuffer>
mt_nativeBuffer(GPUBuffer *buffer) {
  return buffer ? (id<MTLBuffer>)buffer->_priv : nil;
}

GPU_HIDE
GPUComputePipeline*
mt_newComputePipeline(void) {
  GPUComputePipeline *pipeline;
  MTComputePipelineDesc *desc;

  pipeline = calloc(1, sizeof(*pipeline));
  desc = calloc(1, sizeof(*desc));
  if (!pipeline || !desc) {
    free(desc);
    free(pipeline);
    return NULL;
  }

  pipeline->_priv = desc;
  return pipeline;
}

GPU_HIDE
void
mt_setComputeFunction(GPUComputePipeline *pipeline, GPUFunction *func) {
  MTComputePipelineDesc *desc;

  if (!pipeline || !pipeline->_priv || !func) {
    return;
  }

  desc = pipeline->_priv;
  desc->function = (id<MTLFunction>)func->_priv;
}

GPU_HIDE
GPUComputePipelineState*
mt_newComputeState(GPUDevice *device, GPUComputePipeline *pipeline) {
  GPUDeviceMT *deviceMT;
  MTComputePipelineDesc *desc;
  GPUComputePipelineState *state;
  id<MTLComputePipelineState> mtState;
  NSError *error;

  if (!device || !pipeline || !pipeline->_priv) {
    return NULL;
  }

  deviceMT = device->_priv;
  desc = pipeline->_priv;
  if (!deviceMT || !desc->function) {
    return NULL;
  }

  error = nil;
  mtState = [deviceMT->device newComputePipelineStateWithFunction:desc->function
                                                            error:&error];
  if (!mtState) {
    NSLog(@"Failed to create compute pipeline state: %@", error);
    return NULL;
  }

  state = calloc(1, sizeof(*state));
  if (!state) {
    [mtState release];
    return NULL;
  }

  state->_priv = mtState;
  state->workgroupSize[0] = 1u;
  state->workgroupSize[1] = 1u;
  state->workgroupSize[2] = 1u;
  pipeline->_state = state;
  return state;
}

GPU_HIDE
void
mt_destroyComputePipeline(GPUComputePipeline *pipeline) {
  MTComputePipelineDesc *desc;

  if (!pipeline) {
    return;
  }

  desc = pipeline->_priv;
  if (pipeline->_state) {
    GPUComputePipelineState *state = pipeline->_state;
    if (state->_priv) {
      [(id<MTLComputePipelineState>)state->_priv release];
    }
    free(state);
  }
  free(desc);
  free(pipeline);
}

GPU_HIDE
GPUComputePassEncoder*
mt_computeCommandEncoder(GPUCommandBuffer *cmdb, const char *label) {
  MTCommandBuffer       *commandState;
  MTComputeEncoder      *nativeState;
  GPUComputePassEncoder *enc;

  if (!cmdb) {
    return NULL;
  }

  commandState = mt_commandBuffer(cmdb);
  if (!commandState) {
    return NULL;
  }
  enc = &commandState->computeEncoder;
  nativeState = &commandState->computeState;
  memset(enc, 0, sizeof(*enc));
  memset(nativeState, 0, sizeof(*nativeState));

  if (commandState && commandState->mode == MTCommandMode4) {
    if (!mt_prepareArgumentState(cmdb,
                                 &commandState->computeArguments,
                                 gpuDeviceDebugLabel(
                                   gpuCommandBufferDevice(cmdb),
                                   "gpu-metal4-compute-arguments"))) {
      return NULL;
    }
    if (@available(macOS 26.0, iOS 26.0, *)) {
      nativeState->modern = [commandState->modern computeCommandEncoder];
      mt_applyPendingBarrier(cmdb, nativeState->modern);
      nativeState->arguments = &commandState->computeArguments;
      [nativeState->modern setArgumentTable:nativeState->arguments->table];
    }
  } else {
    nativeState->classic = [mt_classicCommandBuffer(cmdb) computeCommandEncoder];
  }

  if (!nativeState->classic && !nativeState->modern) {
    return NULL;
  }
#if GPU_BUILD_WITH_DEBUG_MARKERS
  if (label && label[0] != '\0') {
    NSString *nativeLabel = [NSString stringWithUTF8String:label];

    nativeState->classic.label = nativeLabel;
    if (@available(macOS 26.0, iOS 26.0, *)) {
      [(id<MTL4ComputeCommandEncoder>)nativeState->modern setLabel:nativeLabel];
    }
  }
#else
  GPU__UNUSED(label);
#endif

  enc->_priv = nativeState;
  enc->_workgroupSize[0] = 1u;
  enc->_workgroupSize[1] = 1u;
  enc->_workgroupSize[2] = 1u;
  return enc;
}

GPU_HIDE
void
mt_setComputePipelineState(GPUComputePassEncoder *enc,
                           GPUComputePipelineState *state) {
  MTComputeEncoder *native;

  if (!enc || !state || !state->_priv) {
    return;
  }

  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  if (native && native->modern) {
    if (@available(macOS 26.0, iOS 26.0, *)) {
      [native->modern setComputePipelineState:(id<MTLComputePipelineState>)state->_priv];
    }
  } else {
    [native->classic setComputePipelineState:(id<MTLComputePipelineState>)state->_priv];
  }
  enc->_workgroupSize[0] = state->workgroupSize[0] ? state->workgroupSize[0] : 1u;
  enc->_workgroupSize[1] = state->workgroupSize[1] ? state->workgroupSize[1] : 1u;
  enc->_workgroupSize[2] = state->workgroupSize[2] ? state->workgroupSize[2] : 1u;
}

GPU_HIDE
void
mt_computeBuffer(GPUComputePassEncoder *enc,
                 GPUBuffer             *buf,
                 size_t                 off,
                 uint32_t               index) {
  MTComputeEncoder *native;
  id<MTLBuffer>     buffer;

  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  buffer = mt_nativeBuffer(buf);
  if (native && native->modern) {
    mt_setArgumentBuffer(enc->_cmdb, native->arguments, buffer, off, index);
    return;
  }
  [native->classic setBuffer:buffer offset:(NSUInteger)off atIndex:index];
}

GPU_HIDE
void
mt_computeTexture(GPUComputePassEncoder *enc,
                  GPUTextureView        *view,
                  uint32_t               index) {
  MTComputeEncoder *native;
  id<MTLTexture>    texture;

  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  texture = view ? (id<MTLTexture>)view->_priv : nil;
  if (native && native->modern) {
    mt_setArgumentTexture(enc->_cmdb, native->arguments, texture, index);
    return;
  }
  [native->classic setTexture:texture atIndex:index];
}

GPU_HIDE
void
mt_computeSampler(GPUComputePassEncoder *enc,
                  GPUSampler            *sampler,
                  uint32_t               index) {
  MTComputeEncoder    *native;
  id<MTLSamplerState>  samplerState;

  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  samplerState = sampler ? (id<MTLSamplerState>)sampler->_priv : nil;
  if (native && native->modern) {
    mt_setArgumentSampler(native->arguments, samplerState, index);
    return;
  }
  [native->classic setSamplerState:samplerState atIndex:index];
}

GPU_HIDE
void
mt_computePushConstants(GPUComputePassEncoder *enc,
                        const void            *data,
                        uint32_t               sizeBytes) {
  MTComputeEncoder *native;

  if (!data || sizeBytes == 0u) {
    return;
  }

  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  if (native && native->modern) {
    uint64_t address;

    if (mt_uploadConstants(enc->_cmdb, data, sizeBytes, &address)) {
      if (@available(macOS 26.0, iOS 26.0, *)) {
        [(id<MTL4ArgumentTable>)native->arguments->table
          setAddress:address
             atIndex:MT_PUSH_CONSTANT_INDEX];
      }
      native->arguments->bufferMask |= 1u << MT_PUSH_CONSTANT_INDEX;
    }
    return;
  }
  [native->classic setBytes:data
                     length:(NSUInteger)sizeBytes
                    atIndex:MT_PUSH_CONSTANT_INDEX];
}

GPU_HIDE
void
mt_dispatch(GPUComputePassEncoder *enc,
            uint32_t               x,
            uint32_t               y,
            uint32_t               z) {
  MTComputeEncoder *native;
  MTLSize groups;
  MTLSize threads;

  groups = MTLSizeMake(x, y, z);
  threads = MTLSizeMake(enc->_workgroupSize[0] ? enc->_workgroupSize[0] : 1u,
                        enc->_workgroupSize[1] ? enc->_workgroupSize[1] : 1u,
                        enc->_workgroupSize[2] ? enc->_workgroupSize[2] : 1u);
  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  if (native->modern) {
    if (@available(macOS 26.0, iOS 26.0, *)) {
      [native->modern dispatchThreadgroups:groups
                       threadsPerThreadgroup:threads];
    }
    return;
  }
  [native->classic dispatchThreadgroups:groups threadsPerThreadgroup:threads];
}

GPU_HIDE
void
mt_dispatchIndirect(GPUComputePassEncoder *enc,
                    GPUBuffer             *argsBuffer,
                    uint64_t               argsOffset) {
  MTComputeEncoder *native;
  id<MTLBuffer>     args;
  MTLSize           threads;

  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  args = mt_nativeBuffer(argsBuffer);
  threads = MTLSizeMake(enc->_workgroupSize[0] ? enc->_workgroupSize[0] : 1u,
                        enc->_workgroupSize[1] ? enc->_workgroupSize[1] : 1u,
                        enc->_workgroupSize[2] ? enc->_workgroupSize[2] : 1u);
  if (native && native->modern) {
    if (@available(macOS 26.0, iOS 26.0, *)) {
      mt_useAllocation(enc->_cmdb, args);
      [native->modern dispatchThreadgroupsWithIndirectBuffer:args.gpuAddress + argsOffset
                                       threadsPerThreadgroup:threads];
    }
    return;
  }
  [native->classic dispatchThreadgroupsWithIndirectBuffer:args
                                      indirectBufferOffset:(NSUInteger)argsOffset
                                    threadsPerThreadgroup:threads];
}

GPU_HIDE
void
mt_endComputeEncoding(GPUComputePassEncoder *enc) {
  MTComputeEncoder *native;

  if (!enc) {
    return;
  }
  native = mt_computeEncoder(enc);
  if (!native) {
    return;
  }
  if (native && native->modern) {
    if (@available(macOS 26.0, iOS 26.0, *)) {
      [native->modern endEncoding];
    }
  } else {
    [native->classic endEncoding];
  }
  native->classic = nil;
  native->modern = nil;
}

GPU_HIDE
void
mt_initCompute(GPUApiCompute *api) {
  api->newComputePipeline = mt_newComputePipeline;
  api->setFunction = mt_setComputeFunction;
  api->newComputeState = mt_newComputeState;
  api->destroyComputePipeline = mt_destroyComputePipeline;
  api->computeCommandEncoder = mt_computeCommandEncoder;
  api->setComputePipelineState = mt_setComputePipelineState;
  api->buffer = mt_computeBuffer;
  api->texture = mt_computeTexture;
  api->sampler = mt_computeSampler;
  api->pushConstants = mt_computePushConstants;
  api->dispatch = mt_dispatch;
  api->dispatchIndirect = mt_dispatchIndirect;
  api->endEncoding = mt_endComputeEncoding;
}

File: src/backend/dx12/impl/compute.c
/*
 * Copyright (C) 2026 Recep Aslantas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../common.h"
#include "../impl.h"
#include "../../../api/compute_internal.h"

static GPUComputeEncoderDX12 *
dx12__computeEncoder(GPUComputePassEncoder *encoder) {
  return encoder ? encoder->_priv : NULL;
}

GPU_HIDE
GPUResult
dx12_createComputePipeline(GPUDevice                          *device,
                           const GPUComputePipelineCreateInfo *info,
                           GPUComputePipeline                 *pipeline) {
  GPUDeviceDX12           *deviceDX12;
  GPULibraryDX12          *library;
  GPUPipelineLayoutDX12   *layout;
  GPUComputePipelineState *state;
  GPUComputePipelineDX12  *native;
  ID3D12RootSignature     *rootSignature;
  D3D12_COMPUTE_PIPELINE_STATE_DESC desc = {0};
  DX12ShaderCode           shaderCode = {0};
  HRESULT                  result;

  deviceDX12 = device ? device->_priv : NULL;
  library    = info && info->library ? info->library->_priv : NULL;
  layout     = info && info->layout ? info->layout->_native : NULL;
  if (!deviceDX12 || !deviceDX12->d3dDevice || !library ||
      !library->source || !layout || !layout->rootSignature ||
      !info->entryPoint || !info->entryPoint[0] || !pipeline) {
    return GPU_ERROR_INVALID_ARGUMENT;
  }

  state = calloc(1, sizeof(*state) + sizeof(*native));
  if (!state) {
    return GPU_ERROR_OUT_OF_MEMORY;
  }
  native = (GPUComputePipelineDX12 *)(state + 1);
  rootSignature = NULL;
  if (dx12_createShaderRootSignature(device,
                                     info->layout,
                                     info->library,
                                     &rootSignature) != GPU_OK) {
    free(state);
    return GPU_ERROR_BACKEND_FAILURE;
  }

  if (!dx12_compileShader(deviceDX12,
                          library,
                          info->entryPoint,
                          GPU_SHADER_STAGE_COMPUTE_BIT,
                          &shaderCode)) {
    rootSignature->lpVtbl->Release(rootSignature);
    free(state);
    return GPU_ERROR_BACKEND_FAILURE;
  }

  desc.pRootSignature      = rootSignature;
  desc.CS.pShaderBytecode  = shaderCode.data;
  desc.CS.BytecodeLength   = shaderCode.size;
  result = deviceDX12->d3dDevice->lpVtbl->CreateComputePipelineState(
    deviceDX12->d3dDevice,
    &desc,
    &IID_ID3D12PipelineState,
    (void **)&native->pipelineState
  );
  dx12_freeShaderCode(&shaderCode);
  if (FAILED(result) || !native->pipelineState) {
    rootSignature->lpVtbl->Release(rootSignature);
    free(state);
    return GPU_ERROR_BACKEND_FAILURE;
  }

  native->rootSignature = rootSignature;
  state->_priv            = native;
  state->workgroupSize[0] = 1u;
  state->workgroupSize[1] = 1u;
  state->workgroupSize[2] = 1u;
  pipeline->_priv         = native;
  pipeline->_state        = state;
  return GPU_OK;
}

GPU_HIDE
void
dx12_destroyComputePipeline(GPUComputePipeline *pipeline) {
  GPUComputePipelineState *state;
  GPUComputePipelineDX12  *native;

  if (!pipeline) {
    return;
  }

  state  = pipeline->_state;
  native = state ? state->_priv : NULL;
  if (native) {
    if (native->pipelineState) {
      native->pipelineState->lpVtbl->Release(native->pipelineState);
    }
    if (native->rootSignature) {
      native->rootSignature->lpVtbl->Release(native->rootSignature);
    }
  }
  free(state);
  free(pipeline);
}

GPU_HIDE
GPUComputePassEncoder *
dx12_computeCommandEncoder(GPUCommandBuffer *cmdb, const char *label) {
  GPUDeviceDX12           *device;
  GPUCommandBufferDX12    *command;
  GPUComputePassEncoder   *encoder;
  GPUComputeEncoderDX12   *native;

  device  = cmdb && cmdb->_queue && cmdb->_queue->_device
              ? cmdb->_queue->_device->_priv
              : NULL;
  command = cmdb ? cmdb->_priv : NULL;
  if (!device || !command || !command->commandList) {
    return NULL;
  }

  encoder = &command->computeEncoder;
  native  = &command->computeState;
  memset(encoder, 0, sizeof(*encoder));
  memset(native, 0, sizeof(*native));
  native->device           = device;
  native->commandList      = command->commandList;
  native->debugEventActive = dx12_beginDebugEvent(
    gpuCommandBufferDevice(cmdb),
    native->commandList,
    label
  );
  encoder->_priv             = native;
  encoder->_workgroupSize[0] = 1u;
  encoder->_workgroupSize[1] = 1u;
  encoder->_workgroupSize[2] = 1u;
  return encoder;
}

GPU_HIDE
void
dx12_setComputePipelineState(GPUComputePassEncoder   *encoder,
                             GPUComputePipelineState *pipelineState) {
  GPUComputeEncoderDX12  *native;
  GPUComputePipelineDX12 *pipeline;

  native   = dx12__computeEncoder(encoder);
  pipeline = pipelineState ? pipelineState->_priv : NULL;
  if (!native || !native->commandList || !pipeline ||
      !pipeline->pipelineState || !pipeline->rootSignature) {
    return;
  }

  native->commandList->lpVtbl->SetComputeRootSignature(native->commandList,
                                                       pipeline->rootSignature);
  native->commandList->lpVtbl->SetPipelineState(native->commandList,
                                                pipeline->pipelineState);
  native->rootSignature      = pipeline->rootSignature;
  encoder->_workgroupSize[0] = pipelineState->workgroupSize[0];
  encoder->_workgroupSize[1] = pipelineState->workgroupSize[1];
  encoder->_workgroupSize[2] = pipelineState->workgroupSize[2];
}

GPU_HIDE
void
dx12_computePushConstants(GPUComputePassEncoder *encoder,
                          const void            *data,
                          uint32_t               sizeBytes) {
  GPUComputeEncoderDX12 *native;
  GPUPipelineLayoutDX12 *layout;

  native = dx12__computeEncoder(encoder);
  layout = encoder && encoder->_pipelineLayout
             ? encoder->_pipelineLayout->_native
             : NULL;
  if (!native || !native->commandList || !layout || !data ||
      layout->pushConstantRootParameter == UINT32_MAX ||
      sizeBytes != layout->pushConstantDwordCount * 4u) {
    return;
  }

  native->commandList->lpVtbl->SetComputeRoot32BitConstants(
    native->commandList,
    layout->pushConstantRootParameter,
    layout->pushConstantDwordCount,
    data,
    0u
  );
}

GPU_HIDE
void
dx12_dispatch(GPUComputePassEncoder *encoder,
              uint32_t               x,
              uint32_t               y,
              uint32_t               z) {
  GPUComputeEncoderDX12 *native;

  native = dx12__computeEncoder(encoder);
  if (!native || !native->commandList) {
    return;
  }

  native->commandList->lpVtbl->Dispatch(native->commandList, x, y, z);
}

static bool
dx12__dispatchIndirect(GPUComputePassEncoder *encoder,
                       GPUBuffer             *argsBuffer,
                       uint64_t               argsOffset,
                       uint32_t               dispatchCount,
                       uint32_t               strideBytes) {
  GPUComputeEncoderDX12 *native;
  GPUBufferDX12         *buffer;

  native = dx12__computeEncoder(encoder);
  buffer = argsBuffer ? argsBuffer->_priv : NULL;
  if (!native || !native->device || !native->commandList || !buffer ||
      !buffer->resource || !native->device->dispatchSignature ||
      !dx12_transitionBuffer(native->commandList,
                             buffer,
                             D3D12_RESOURCE_STATE_INDIRECT_ARGUMENT) ||
      strideBytes != (uint32_t)sizeof(D3D12_DISPATCH_ARGUMENTS)) {
    return false;
  }

  native->commandList->lpVtbl->ExecuteIndirect(
    native->commandList,
    native->device->dispatchSignature,
    dispatchCount,
    buffer->resource,
    argsOffset,
    NULL,
    0u
  );
  return true;
}

GPU_HIDE
void
dx12_dispatchIndirect(GPUComputePassEncoder *encoder,
                      GPUBuffer             *argsBuffer,
                      uint64_t               argsOffset) {
  (void)dx12__dispatchIndirect(encoder,
                               argsBuffer,
                               argsOffset,
                               1u,
                               (uint32_t)sizeof(D3D12_DISPATCH_ARGUMENTS));
}

GPU_HIDE
bool
dx12_multiDispatchIndirect(GPUComputePassEncoder *encoder,
                           GPUBuffer             *argsBuffer,
                           uint64_t               argsOffset,
                           uint32_t               dispatchCount,
                           uint32_t               strideBytes) {
  return dx12__dispatchIndirect(encoder,
                                argsBuffer,
                                argsOffset,
                                dispatchCount,
                                strideBytes);
}

GPU_HIDE
void
dx12_endComputeEncoding(GPUComputePassEncoder *encoder) {
  GPUComputeEncoderDX12 *native;

  native = dx12__computeEncoder(encoder);
  if (!native) {
    return;
  }

  if (native->debugEventActive) {
    dx12_endDebugEvent(gpuCommandBufferDevice(encoder->_cmdb),
                       native->commandList);
  }

  native->device           = NULL;
  native->commandList      = NULL;
  native->rootSignature    = NULL;
  native->resourceHeap     = NULL;
  native->samplerHeap      = NULL;
  native->debugEventActive = false;
}

GPU_HIDE
void
dx12_initCompute(GPUApiCompute *api) {
  api->createPipeline          = dx12_createComputePipeline;
  api->destroyComputePipeline  = dx12_destroyComputePipeline;
  api->computeCommandEncoder   = dx12_computeCommandEncoder;
  api->setComputePipelineState = dx12_setComputePipelineState;
  api->pushConstants           = dx12_computePushConstants;
  api->dispatch                = dx12_dispatch;
  api->dispatchIndirect        = dx12_dispatchIndirect;
  api->multiDispatchIndirect   = dx12_multiDispatchIndirect;
  api->endEncoding             = dx12_endComputeEncoding;
}
